tests/test_grammar.py

```diff
@@ -250,7 +280,7 @@
     ("nested_select", "${options.${choice}}", "A"),
     ("nested_select_getitem", "${options[${choice}]}", "A"),
     ("nested_relative", "${${rel_opt}.b}", "B"),
-    ("str_quoted_nested", r"'AB${test:\'CD${test:\\'EF\\'}GH\'}'", "ABCDEFGH"),
+    ("str_quoted_nested", r"'AB${test:\'CD${test:\\\'EF\\\'}GH\'}'", "ABCDEFGH"),
```

Comment:
Can you explain this one?
Is this something that will be addressed by the followup double escaping diff?

Reply:
No, I would expect this one to remain unchanged. Here's what happens:
- Un-escaping of quoted string: `AB${test:\'CD${test:\\\'EF\\\'}GH\'}` ==> `AB${test:'CD${test:\'EF\'}GH'}`
- Un-escaping of top-level strings: no change because the only top-level string is `AB`, the rest is an interpolation and interpolations aren't touched in this step
- We then start processing the interpolation
- Un-escaping of quoted string: `CD${test:\'EF\'}GH` ==> `CD${test:'EF'}GH`
- Un-escaping of top-level strings: no change (the top-level strings are `CD` and `GH`)
- We then process the interpolation which is a normal one (yielding the string `EF`)

The reason why it worked before with *fewer* backslashes in the middle is because a string like `'foo\\'bar'` was accepted as a valid quoted string by the grammar, while this is not the case anymore: now this would be parsed as a quoted string `'foo\\'` followed by `bar'` which is an invalid expression. I don't think we could have kept the previous behavior without facing the subtle bugs of #617.

Reply:
1. So the reason we are unescaping in top level strings is just to support escaping interpolations there?
If so, would it simplify things to just say that we only apply the grammar inside interpolations and thus only support escaping of nested interpolations?
2. I am a bit confused by this:
```
input:  r"'AB${test:\'CD${test:\\'EF\\'}GH\'}'"
actual string: 'AB${test:\'CD${test:\\'EF\\'}GH\'}'
```

This is a string outside of interpolation, and it's quoted. why is the result not quoted as well?
I am assuming that this string is interpreted as a value of a config node, so there is no yaml involved here.

3. I don't like how we need to unescape the nested quoted strings.
feels like it adds a lot of complexity to the mental model required to understand the behavior of quoted strings. before diving into this one, I want to understand 2.

Reply:
> 1. So the reason we are unescaping in top level strings is just to support escaping interpolations there?

Yes.

>    If so, would it simplify things to just say that we only apply the grammar inside interpolations and thus only support escaping of nested interpolations?

I may misunderstand what you're suggesting, but how would you set a config option that evaluates to the string with characters `Please use ${VAR}` if we can't escape interpolations in top-level strings?

> 2. I am a bit confused by this:
> 
> ```
> input:  r"'AB${test:\'CD${test:\\'EF\\'}GH\'}'"
> actual string: 'AB${test:\'CD${test:\\'EF\\'}GH\'}'
> ```
> 
> This is a string outside of interpolation, and it's quoted. why is the result not quoted as well?
> I am assuming that this string is interpreted as a value of a config node, so there is no yaml involved here.

Oh! Careful, there are several tests in the file, and this one belongs to the tests for the `singleElement` parser rule (not `configValue`), so this is parsed as a quoted string like you would use as a resolver argument.


> 3. I don't like how we need to unescape the nested quoted strings.
>    feels like it adds a lot of complexity to the mental model required to understand the behavior of quoted strings. before diving into this one, I want to understand 2.

Let me know how you feel now that hopefully point 2 is clearer.

Reply:
> I may misunderstand what you're suggesting, but how would you set a config option that evaluates to the string with characters Please use ${VAR} if we can't escape interpolations in top-level strings?

How would you do it in OmegaConf 2.0 (not supported at all as far as I can tell).

In principle, you could use decode (although I think it's not working right now for this)
```
a: ${oc.decode:'Please use \${VAR}'}
```

> Oh! Careful, there are several tests in the file, and this one belongs to the tests for the singleElement parser rule (not configValue), so this is parsed as a quoted string like you would use as a resolver argument.

This is one of the reasons I don't like the current testing approach: it makes it hard to understand what the test data is being used for.

> Let me know how you feel now that hopefully point 2 is clearer.

Okay, so back to nested quoted strings:
```
outer quoted string: 
'AB${test:\'CD${test:\\'EF\\'}GH\'}'
```
At some point we considered using lexer modes to support something like this:
```
'AB${test:'CD${test:'EF'}GH'}'
```
(Correct me if I am wrong and lexer modes will not actually enable this).
This is becoming much more attractive after the back and forth we had on this issue.
Can we pull this off?
With the exception of the time investment, do you see any downsides to suspect additional corner cases?

Reply:
> > I may misunderstand what you're suggesting, but how would you set a config option that evaluates to the string with characters Please use ${VAR} if we can't escape interpolations in top-level strings?
> 
> How would you do it in OmegaConf 2.0 (not supported at all as far as I can tell).

Indeed, that was a gap in 2.0. Though you could use `${}` as long as it didn't look like an interpolation, for instance `"Please use ${VAR!}"` would work.

> In principle, you could use decode (although I think it's not working right now for this)
> 
> ```
> a: ${oc.decode:'Please use \${VAR}'}
> ```

The example you provide doesn't work because the string that is given as input to `oc.decode` has characters `Please use ${VAR}` so it crashes trying to resolve the interpolation (in `oc.decode`).

I'm actually surprised that it's allowed by the grammar because I didn't realize that you could have interpolations in unquoted strings (`oc.decode` uses the `singleElement` parser rule). So for instance you can write this:
```python
cfg = OmegaConf.create({"x": "foo", "y": "${identity:abc ${x}}"})
assert cfg.y == "abc foo"
```
I'm not sure this is even tested right now, I'll double check later.

This works with `oc.decode`:
```python
cfg = OmegaConf.create({"x": r"""${oc.decode:'"Please use \\\\\${VAR}"'}"""})
assert cfg.x == "Please use ${VAR}"
```

That being said, even if we decide that this is how to do it, it still relies on the escaping of interpolations at the top-level, because the parsing of a quoted string triggers a top-level parsing after the removal of quotes & un-escaping. So I'm not sure what we would gain.

> Okay, so back to nested quoted strings:
> 
> ```
> outer quoted string: 
> 'AB${test:\'CD${test:\\'EF\\'}GH\'}'
> ```
> 
> At some point we considered using lexer modes to support something like this:
> 
> ```
> 'AB${test:'CD${test:'EF'}GH'}'
> ```
> 
> (Correct me if I am wrong and lexer modes will not actually enable this).
> This is becoming much more attractive after the back and forth we had on this issue.
> Can we pull this off?
> With the exception of the time investment, do you see any downsides to suspect additional corner cases?

Almost -- you actually need to alternate quote styles to make it work, otherwise there is no way to differentiate (in the greedy way the lexer works) a closing quote vs. one that starts a new segment. So it should instead be:
```
 'AB${test:"CD${test:'EF'}GH"}'
```
I think it should work fine (it used to), just flagging that it means we would now parse the content of quoted strings when we parse an expression (not really a problem, just to make it clear), instead of triggering this parsing only on evaluation.
The main downside I see is the one you had raised initially: it may be an unusual approach that could surprise people.

Reply:
> > > I may misunderstand what you're suggesting, but how would you set a config option that evaluates to the string with characters Please use ${VAR} if we can't escape interpolations in top-level strings?
> > 
> > 
> > How would you do it in OmegaConf 2.0 (not supported at all as far as I can tell).
> 
> Indeed, that was a gap in 2.0. Though you could use `${}` as long as it didn't look like an interpolation, for instance `"Please use ${VAR!}"` would work.

This is one of many enhancements in 2.1, but it's not even a requested one so I would be okay with an alternative implementation if it's reasonably clean and simplify out lives.

> > In principle, you could use decode (although I think it's not working right now for this)
> > ```
> > a: ${oc.decode:'Please use \${VAR}'}
> > ```
> 
> The example you provide doesn't work because the string that is given as input to `oc.decode` has characters `Please use ${VAR}` so it crashes trying to resolve the interpolation (in `oc.decode`).

This is a feature of oc.decode:
1. I am not sure why it should resolve the interpolation in the first place.
2. the interpolation is escaped, so it should understand that and not try to resolve it (in principle, I am not talking about the current implementation).

> I'm actually surprised that it's allowed by the grammar because I didn't realize that you could have interpolations in unquoted strings (`oc.decode` uses the `singleElement` parser rule). So for instance you can write this:

I used a quoted string, so I am not sure why you were surprised.

> ```python
> cfg = OmegaConf.create({"x": "foo", "y": "${identity:abc ${x}}"})
> assert cfg.y == "abc foo"
> ```
> 
> I'm not sure this is even tested right now, I'll double check later.

I have done it before, I usually quote just for clarity in scenarios where the string contains spaces.
Let me know if you find why this works :).

> This works with `oc.decode`:
> 
> ```python
> cfg = OmegaConf.create({"x": r"""${oc.decode:'"Please use \\\\\${VAR}"'}"""})
> assert cfg.x == "Please use ${VAR}"
> ```
> 
> That being said, even if we decide that this is how to do it, it still relies on the escaping of interpolations at the top-level, because the parsing of a quoted string triggers a top-level parsing after the removal of quotes & un-escaping. So I'm not sure what we would gain.

Obviously this is not great. but if we can make `a: ${oc.decode:'Please use \${VAR}'}` work I would vote to simplify top-level string and maybe even drop it completely.

> > Okay, so back to nested quoted strings:
> > ```
> > outer quoted string: 
> > 'AB${test:\'CD${test:\\'EF\\'}GH\'}'
> > ```
> > 
> > 
> > At some point we considered using lexer modes to support something like this:
> > ```
> > 'AB${test:'CD${test:'EF'}GH'}'
> > ```
> > 
> > 
> > (Correct me if I am wrong and lexer modes will not actually enable this).
> > This is becoming much more attractive after the back and forth we had on this issue.
> > Can we pull this off?
> > With the exception of the time investment, do you see any downsides to suspect additional corner cases?
> 
> Almost -- you actually need to alternate quote styles to make it work, otherwise there is no way to differentiate (in the greedy way the lexer works) a closing quote vs. one that starts a new segment. So it should instead be:
> 
> ```
>  'AB${test:"CD${test:'EF'}GH"}'
> ```

I am not sure there is no way to differentiate:
Once we are in a quoted string (say single), hitting another single quote closes it.
However, if we hit a double quote we are good, but if we hit an interpolation token (IE, non-greedy), 
we can push a mode that allows us to start a new single quoted string.
when we seen an interpolation end, we pop (By the way, this might make it impossible for regex to handle with the grammar).

> I think it should work fine (it used to), just flagging that it means we would now parse the content of quoted strings when we parse an expression (not really a problem, just to make it clear), instead of triggering this parsing only on evaluation.
> The main downside I see is the one you had raised initially: it may be an unusual approach that could surprise people.


After seeing what where the alternative path leads, I am much more open to a bit of weirdness in exchange for something that is much easier to grok.

Reply:
> > > In principle, you could use decode (although I think it's not working right now for this)
> > > ```
> > > a: ${oc.decode:'Please use \${VAR}'}
> > > ```
> > 
> > 
> > The example you provide doesn't work because the string that is given as input to `oc.decode` has characters `Please use ${VAR}` so it crashes trying to resolve the interpolation (in `oc.decode`).
> 
> This is a feature of oc.decode:
> 
> 1. I am not sure why it should resolve the interpolation in the first place.
> 2. the interpolation is escaped, so it should understand that and not try to resolve it (in principle, I am not talking about the current implementation).

1. `oc.decode` currently relies on the same parsing (& visitor) code as the one used to parse (& resolve) the `singleElement` rule of the grammar, which resolves interpolations.

2. But `oc.decode` doesn't see the escaped interpolation: it sees the string with characters `Please use ${VAR}` where the interpolation isn't escaped. The reason is that the quoted string `'Please use \${VAR}'` is first resolved into `Please use \${VAR}` *then* passed to `oc.decode`. So what you're trying to achieve would be done in a simpler way with an `identity` resolver that would return this string unchanged (instead of `oc.decode` which parses it according to the grammar rules).


> > I'm actually surprised that it's allowed by the grammar because I didn't realize that you could have interpolations in unquoted strings (`oc.decode` uses the `singleElement` parser rule). So for instance you can write this:
> 
> I used a quoted string, so I am not sure why you were surprised.

Hopefully the above comment makes it clearer: `oc.decode` sees the string with characters `Please use ${VAR}` which is unquoted.

> > ```python
> > cfg = OmegaConf.create({"x": "foo", "y": "${identity:abc ${x}}"})
> > assert cfg.y == "abc foo"
> > ```
> > 
> > 
> > I'm not sure this is even tested right now, I'll double check later.
> 
> I have done it before, I usually quote just for clarity in scenarios where the string contains spaces.
> Let me know if you find why this works :)

Oh, a quick look at the grammar makes it clear why it works. I was just surprised because for some reason I thought it wouldn't. I just want to make sure it's properly tested.

> > This works with `oc.decode`:
> > ```python
> > cfg = OmegaConf.create({"x": r"""${oc.decode:'"Please use \\\\\${VAR}"'}"""})
> > assert cfg.x == "Please use ${VAR}"
> > ```
> > 
> > 
> > That being said, even if we decide that this is how to do it, it still relies on the escaping of interpolations at the top-level, because the parsing of a quoted string triggers a top-level parsing after the removal of quotes & un-escaping. So I'm not sure what we would gain.
> 
> Obviously this is not great. but if we can make `a: ${oc.decode:'Please use \${VAR}'}` work I would vote to simplify top-level string and maybe even drop it completely.

I don't think we can make this one work (for the reason explained above), but if we remove the un-escaping of `\\` from quoted strings (as in #621) then we can probably shave off a few `\` from my example.
I'm not really sure how we would "simplify top-level string and maybe even drop it completely", since the parsing of quoted strings relies on it as well.

> > > At some point we considered using lexer modes to support something like this:
> > > ```
> > > 'AB${test:'CD${test:'EF'}GH'}'
> > > ```
> > > 
> > > 
> > > Can we pull this off?
> > 
> > Almost -- you actually need to alternate quote styles to make it work, otherwise there is no way to differentiate (in the greedy way the lexer works) a closing quote vs. one that starts a new segment. So it should instead be:
> > ```
> >  'AB${test:"CD${test:'EF'}GH"}'
> > ```
> 
> I am not sure there is no way to differentiate:
> Once we are in a quoted string (say single), hitting another single quote closes it.
> However, if we hit a double quote we are good, but if we hit an interpolation token (IE, non-greedy),
> we can push a mode that allows us to start a new single quoted string.
> when we seen an interpolation end, we pop (By the way, this might make it impossible for regex to handle with the grammar).

Oh yes you are right, thanks! I'm not sure where this idea of alternating came from -- for some reason I thought I remembered you had to alternate, but I think I just imagined it.

> After seeing what where the alternative path leads, I am much more open to a bit of weirdness in exchange for something that is much easier to grok.

Alright, I'm happy to do it, but then we also need to decide if we still want to support escaped quotes in quoted strings. I *think* we don't need them anymore because we should always be able to use a different enclosing quote style instead, for instance `"there is a single quote ' in the middle"`. I haven't thought in depth about it though.

Reply:
> 2. But `oc.decode` doesn't see the escaped interpolation: it sees the string with characters `Please use ${VAR}` where the interpolation isn't escaped. The reason is that the quoted string `'Please use \${VAR}'` is first resolved into `Please use \${VAR}` _then_ passed to `oc.decode`. So what you're trying to achieve would be done in a simpler way with an `identity` resolver that would return this string unchanged (instead of `oc.decode` which parses it according to the grammar rules).

The line of reasoning here seems wrong or missing a step.
But in any case, we shouldn't force decode to do what its' not meant to do.
IF identity works, we can introduce it under some name that makes sense. The use case for not resolving interpolations is very rare and I don't mind if it takes a bit of extra work.

> > Obviously this is not great. but if we can make `a: ${oc.decode:'Please use \${VAR}'}` work I would vote to simplify top-level string and maybe even drop it completely.
> 
> I don't think we can make this one work (for the reason explained above), but if we remove the un-escaping of `\\` from quoted strings (as in #621) then we can probably shave off a few `\` from my example.
> I'm not really sure how we would "simplify top-level string and maybe even drop it completely", since the parsing of quoted strings relies on it as well.

Can you explain what you mean by "since the parsing of quoted strings relies on it as well"?
As far as I understand, quoted strings are only special inside interpolations.
Outside interpolations we can have whatever we want, including quotes.

> > > > At some point we considered using lexer modes to support something like this:
> > > > ```
> > > > 'AB${test:'CD${test:'EF'}GH'}'
> > > > ```
> > > > Can we pull this off?
> > > 
> > > Almost -- you actually need to alternate quote styles to make it work, otherwise there is no way to differentiate (in the greedy way the lexer works) a closing quote vs. one that starts a new segment. So it should instead be:
> > > ```
> > >  'AB${test:"CD${test:'EF'}GH"}'
> > > ```
> > 
> > 
> > I am not sure there is no way to differentiate:
> > Once we are in a quoted string (say single), hitting another single quote closes it.
> > However, if we hit a double quote we are good, but if we hit an interpolation token (IE, non-greedy),
> > we can push a mode that allows us to start a new single quoted string.
> > when we seen an interpolation end, we pop (By the way, this might make it impossible for regex to handle with the grammar).
> 
> Oh yes you are right, thanks! I'm not sure where this idea of alternating came from -- for some reason I thought I remembered you had to alternate, but I think I just imagined it.

Maybe alternating was something we considered before we found out about lexer modes.

> > After seeing what where the alternative path leads, I am much more open to a bit of weirdness in exchange for something that is much easier to grok.
> 
> Alright, I'm happy to do it,

Great, so just to make sure I understand:
In the case of something like:
`${foo: 'outer ${bar: 'inner' }' }`

To be clear, this is the inner interpolation: `${bar: 'inner' }`
It lives in the quotes string `'outer ${bar: 'inner' }'`,
which is a parameter of the resolver `${foo: 'outer ${bar: 'inner' }' }`

What would the tokens be be?
Does it mean that we are promoting quoted string from a token to a parser rule?

> but then we also need to decide if we still want to support escaped quotes in quoted strings. I _think_ we don't need them anymore because we should always be able to use a different enclosing quote style instead, for instance `"there is a single quote ' in the middle"`. I haven't thought in depth about it though.

I think we still need those: how would you create a quoted string that contains both quotes?
```
single: ', double: "
```

Reply:
In any case, before you start working on it, what do you think about merging what we have now? It will be easier starting from a clean slate.

Reply:
Just to follow-up on:

> Oh, a quick look at the grammar makes it clear why it works. I was just surprised because for some reason I thought it wouldn't. I just want to make sure it's properly tested.

==> It's indeed properly tested (there is a section "Interpolations in unquoted strings"), so no worries there (except for my failing memory).

> > 1. But `oc.decode` doesn't see the escaped interpolation: it sees the string with characters `Please use ${VAR}` where the interpolation isn't escaped. The reason is that the quoted string `'Please use \${VAR}'` is first resolved into `Please use \${VAR}` _then_ passed to `oc.decode`. So what you're trying to achieve would be done in a simpler way with an `identity` resolver that would return this string unchanged (instead of `oc.decode` which parses it according to the grammar rules).
> 
> The line of reasoning here seems wrong or missing a step.

It may become clearer with a simpler example:
```python
cfg = OmegaConf.create({"x": "foo", "y": "${identity:'x is ${x}'}"})
cfg.y  # 'x is foo'
```
What happens is that in the visitor, when we are to resolve `${identity:'x is ${x}'}` in `visitInterpolationResolver()`, we first obtain the name of the resolver (`identity`) then obtain the list of arguments to the resolver from `visitSequence()`, which calls `visitElement()` on each argument. The quoted value `'x is ${x}'` is a primitive and its resolution triggers a new parsing pass over its content (`x is ${x}`) where the interpolation is resolved, resulting in the string `x is foo`. *Then* this string is passed to the resolver function associated to `identity`, which returns it unchanged.

The main point here is that the quoted string is evaluted *before* it is passed to the resolver. So, now, if you escape the interpolation:
```python
cfg = OmegaConf.create({"x": "foo", "y": r"${identity:'x is \${x}'}"})
cfg.y  # 'x is ${x}'
```
==> the exact same steps are applied, except that the resolution of the quoted value `'x is \${x}'` results in the string `x is ${x}` which is passed as input to the resolver function associated to `identity`.

Hopefully this makes it clearer why `oc.deode` sees the string `Please use ${VAR}` as input in our previous example.

> > > Obviously this is not great. but if we can make `a: ${oc.decode:'Please use \${VAR}'}` work I would vote to simplify top-level string and maybe even drop it completely.
> > 
> > 
> > I don't think we can make this one work (for the reason explained above), but if we remove the un-escaping of `\\` from quoted strings (as in #621) then we can probably shave off a few `\` from my example.
> > I'm not really sure how we would "simplify top-level string and maybe even drop it completely", since the parsing of quoted strings relies on it as well.
> 
> Can you explain what you mean by "since the parsing of quoted strings relies on it as well"?
> As far as I understand, quoted strings are only special inside interpolations.
> Outside interpolations we can have whatever we want, including quotes.

What I mean is that the resolution of a quoted string currently works as follows:
1. Remove the enclosing quotes (let's say these are double quotes `"`)
2. Un-escape `\\` and `\"` (*not* escaped interpolations)
3. Trigger a new parsing + resolution of the resulting string with the `configValue` rule (which relies on top-level strings)

This will change with the changes discussed below though, but maybe we can discuss it in another PR related to these changes.

> Maybe alternating was something we considered before we found out about lexer modes.


> Great, so just to make sure I understand:
> In the case of something like:
> `${foo: 'outer ${bar: 'inner' }' }`
> 
> To be clear, this is the inner interpolation: `${bar: 'inner' }`
> It lives in the quotes string `'outer ${bar: 'inner' }'`,
> which is a parameter of the resolver `${foo: 'outer ${bar: 'inner' }' }`
> 
> What would the tokens be be?
> Does it mean that we are promoting quoted string from a token to a parser rule?

Yes.
I have a backup of my old branch which should make things clear:
Lexer: https://github.com/odelalleau/omegaconf/blob/backup_2020_08_09/omegaconf/grammar/InterpolationLexer.g4
Parser: https://github.com/odelalleau/omegaconf/blob/backup_2020_08_09/omegaconf/grammar/InterpolationParser.g4

> > but then we also need to decide if we still want to support escaped quotes in quoted strings. I _think_ we don't need them anymore because we should always be able to use a different enclosing quote style instead, for instance `"there is a single quote ' in the middle"`. I haven't thought in depth about it though.
> 
> I think we still need those: how would you create a quoted string that contains both quotes?

Yes indeed, I thought of the same example afterwards. Although I'm fine with keeping escaped quotes, just note that since it's a pretty rare situation we could also instead use resolvers (and we could do the same for escaped interpolations). Something like:
```python
"single ' quote, double ${oc.esc.quote:double} quote, escaped ${oc.esc.interpolation:open} interpolation }"
```
which would be resolved into `single ' quote, double " quote, escaped ${ interpolation }`.
(just throwing it out there as an idea, I don't feel strongly either way)

> In any case, before you start working on it, what do you think about merging what we have now? It will be easier starting from a clean slate.

Agreed.

Reply:
> What I mean is that the resolution of a quoted string currently works as follows:
> 
> 1. Remove the enclosing quotes (let's say these are double quotes `"`)
> 2. Un-escape `\\` and `\"` (_not_ escaped interpolations)
> 3. Trigger a new parsing + resolution of the resulting string with the `configValue` rule (which relies on top-level strings)
> 
> This will change with the changes discussed below though, but maybe we can discuss it in another PR related to these changes.

Yes, it will be easier to discuss in context, and we are considering implementation changes to this anyway.

> > Does it mean that we are promoting quoted string from a token to a parser rule?
> 
> Yes.
> I have a backup of my old branch which should make things clear:
> Lexer: https://github.com/odelalleau/omegaconf/blob/backup_2020_08_09/omegaconf/grammar/InterpolationLexer.g4
> Parser: https://github.com/odelalleau/omegaconf/blob/backup_2020_08_09/omegaconf/grammar/InterpolationParser.g4

Great.

> > > but then we also need to decide if we still want to support escaped quotes in quoted strings. I _think_ we don't need them anymore because we should always be able to use a different enclosing quote style instead, for instance `"there is a single quote ' in the middle"`. I haven't thought in depth about it though.
> > I think we still need those: how would you create a quoted string that contains both quotes?
> Yes indeed, I thought of the same example afterwards. Although I'm fine with keeping escaped quotes, just note that since it's a pretty rare situation we could also instead use resolvers (and we could do the same for escaped interpolations). Something like:
> 
> ```python
> "single ' quote, double ${oc.esc.quote:double} quote, escaped ${oc.esc.interpolation:open} interpolation }"
> ```

I agree that it's a rare case. but the standard approach for it (in most grammars) is to escape and it might be worth supporting directly (people would intuitively try to escape a quote like that).
My gut feeling is that if we switch to the proposed approach, the rare cases where we need to escape will be much easier to understand, so they would no longer be a concern.

> > In any case, before you start working on it, what do you think about merging what we have now? It will be easier starting from a clean slate.
> 
> Agreed.

Will make a final pass soon, not sure I will have time today (Friday is when I have the most meetings).